File: go/hashtable/leetcode/array_intersection_ii.go
package leetcode

import (
	"sort"
)

//求两个数组的交集II

//使用一个map来计数
func intersect(nums1 []int, nums2 []int) []int {
	if len(nums1) > len(nums2) {
		return intersect(nums2, nums1)
	}

	m := map[int]int{}
	for _, v := range nums1 {
		m[v] = m[v] + 1
	}

	var res []int
	for _, v := range nums2 {
		if m[v] > 0 {
			res = append(res, v)
			m[v] = m[v] - 1
		}
	}
	return res
}

//排序之后，使用双指针
func intersect2(nums1 []int, nums2 []int) []int {
	sort.Ints(nums1)
	sort.Ints(nums2)

	i, j := 0, 0
	var res []int
	for i < len(nums1) && j < len(nums2) {
		if nums1[i] == nums2[j] {
			res = append(res, nums1[i])
			i++
			j++
		} else if nums1[i] < nums2[j] {
			i++
		} else {
			j++
		}
	}
	return res
}

File: python/hashtable/leetcode/array_intersection_ii.py

# 两个数组的交集II
from typing import List


class Solution:

    # 使用一个哈希表计数
    def intersect_1(self, nums1: List[int], nums2: List[int]) -> List[int]:
        m, res = {}, []
        for n in nums1:
            m[n] = m[n] + 1 if n in m else 1
        for n in nums2:
            if n in m and m[n] > 0:
                res.append(n)
                m[n] -= 1
        return res

    # 排序之后，使用双指针
    def intersect_2(self, nums1: List[int], nums2: List[int]) -> List[int]:
        nums1.sort(), nums2.sort()
        res, i, j = [], 0, 0
        while i < len(nums1) and j < len(nums2):
            if nums1[i] < nums2[j]:
                i += 1
            elif nums1[i] > nums2[j]:
                j += 1
            else:
                res.append(nums1[i])
                i, j = i + 1, j + 1
        return res
